### PR TITLE
GEODE-7902: fix flaky test

### PR DESCRIPTION
**Pull request by jinmeiliao**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by jinmeiliao** [comment]
> I don't understand this fix.
> 
> 1. Why do we have to sleep between stopping and restarting the server?
> 2. What does it mean for the new jars to "settle" on the file server?
> 3. If the version was correct before the server stopped, how did it revert to an earlier version after it was restarted?
> 
> Also, I'm hoping for a fix that does not involve a fixed wait. If we didn't do that fixed wait at this spot, would the server eventually pick up the correct version of the jar? If not, why not?

To be honest, I don't quite understand either, but after adding the wait, the test ran more than 100 times without failure, before that, it failed at the 2nd time.

File: geode-assembly/src/distributedTest/java/org/apache/geode/management/internal/rest/DeploymentManagementRedployDUnitTest.java [hunk 1]
```diff
@@ -134,14 +135,16 @@ public void redeployJarsWithNewVersionsOfFunctionsAndMultipleLocators() {
...
 
-
     deployment.setFile(jarAVersion2);
     assertManagementResult(client.create(deployment)).isSuccessful();
     server.invoke(() -> assertThatCanLoad(JAR_NAME_A, FUNCTION_A));
     server.invoke(() -> assertThatFunctionHasVersion(FUNCTION_A, VERSION2));
 
     server.stop(false);
 
+    // GEODE-7902: wait for a while for new jars to settle in the file system
```
Comment (dschneider-pivotal) [inline geode-assembly/src/distributedTest/java/org/apache/geode/management/internal/rest/DeploymentManagementRedployDUnitTest.java]: What is the sleep waiting for? What does "settle" mean? Seems like whatever it means couldn't we then do a file system await to figure out if the file has settled?

Comment (demery-pivotal) [reply geode-assembly/src/distributedTest/java/org/apache/geode/management/internal/rest/DeploymentManagementRedployDUnitTest.java]: Would an `awaitUntilAsserted()` around the final assertion (or the final two) suffice, or is it necessary to wait (for whatever the condition is) before doing the restart?

If `awaitUntilAsserted()` works, that's a much more robust fix than a fixed-duration sleep.

Comment (jinmeiliao) [reply geode-assembly/src/distributedTest/java/org/apache/geode/management/internal/rest/DeploymentManagementRedployDUnitTest.java]: I believe the wait has to be before the server restarts to get the new version of jars from one of the locators.

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/functions/DownloadJarFunction.java [hunk 2]
```diff
@@ -61,6 +61,13 @@ public void execute(FunctionContext<Object[]> context) {
         RemoteInputStreamServer istream = null;
         try {
           File jarFile = sharedConfig.getPathToJarOnThisLocator(group, jarName).toFile();
+          // if file is modified within the last second, wait for a second before downloading it
+          // It looks like if the file was modified within the same second, the file downloaded
+          // will be the old content. On mac/linux, the lastModified time of the file is already
+          // reported to the precision of seconds.
+          if (jarFile.lastModified() > System.currentTimeMillis() - 1000) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/management/internal/configuration/functions/DownloadJarFunction.java]: This should be replaced with an await so we don't get stuck here forever.

Comment (jinmeiliao) [reply geode-core/src/main/java/org/apache/geode/management/internal/configuration/functions/DownloadJarFunction.java]: I tried that but await isn't available in core.

**Comment by kirklund** [comment]
Outside of test src sets, you can code your own await. Take a look at geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockService.java's sleep method.

Maybe change it to just throw InterruptedException and add a condition (for example, if want to wait until the pidFile disappears from the JVM's working directory):
```
  private void awaitPidFileDeletion(long millis) throws InterruptedException {
    StopWatch timer = new StopWatch(true);
    while (pidFile.exists()) {
      long timeLeft = millis - timer.elapsedTimeMillis();
      if (timeLeft <= 0) {
        break;
      }
      Thread.sleep(timeLeft);
      break;
    }
  }
```

**Comment by jinmeiliao** [comment]
> Outside of test src sets, you can code your own await. Take a look at geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockService.java's sleep method.
> 
> Maybe change it to just throw InterruptedException and add a condition (for example, if want to wait until the pidFile disappears from the JVM's working directory):
> 
> ```
>   private void awaitPidFileDeletion(long millis) throws InterruptedException {
>     StopWatch timer = new StopWatch(true);
>     while (pidFile.exists()) {
>       long timeLeft = millis - timer.elapsedTimeMillis();
>       if (timeLeft <= 0) {
>         break;
>       }
>       Thread.sleep(timeLeft);
>       break;
>     }
>   }
> ```

Thanks! Decided to put the wait in the test code. But this is helpful.

**Pull request merged**